docs/rfc/rfc-022-command-line-tools.md

```diff
@@ -0,0 +1,187 @@
+unpredictable ways with changes across different releases because of other
+internal plumbing changes.  For example, Tendermint v0.35.x includes changes to
+the way the node configuration file is discovered, and that change affects some
+of the utility subcommands. We verified they worked for v0.35.x, but when we
+backported some subcommand changes into v0.34.x, there were some hidden
+dependencies that weren't included, leading to [#8369][issue8369].
+
+Historically, the argument for including these subcommands into the node binary
+was to avoid making node operators download a lot of individual separate tools
+for important but specialized use cases. This is a reasonable goal, but should
+also be balanced against the goal of having a clear, robust, well-defined, and
+reliable command-line API for the consensus node.
+
+In light of the above, I propose we should take a more deliberate approach to
+the design of the `tendermint` command-line API, one that makes a more clear
+separation between essential functionality and specialized utilities. The
+sections below discuss two specific approaches we could take.
+
+### Approach 1: Command Groups
+
+One way to improve the API is to make more extensive use of command "groups".
+We have already done this to some extent in v0.36 with the `reset` group, which
+collects `reset-state`, `unsafe-reset-all`, and `unsafe-reset-priv-validator`
+under a single heading.
+
+Advantages:
+
+- Grouping "dangerous" or special-purpose commands under a heading communicates
+  intent more clearly.
+
+- Adding additional groups is relatively easy to do, and operators would have
+  to update their scripts but would not need to download or install any new
+  tools.
+
+Disadvantages:
+
+- Fixes to utility commands are still connected tightly to Tendermint releases,
+  making it difficult to hotfix subcommand bugs and provide quick remediation
+  for issues that operators are facing without tagging a whole new release.
+  This leads to new releases including half-baked CLI addons that wind up as a
+  maintenance headache for both the TM Core team and operators later on.
+
+- Libraries like the SDK and tools like `gaiad` that wrap Tendermint commands
+  can easily export specialized, dangerous, and potentially non-production
+  ready functionality to higher-level consumers without good signaling.
+
+### Approach 2: Utility Tool
+
+Another approach would be to split some of the "utility" subcommands into a
+separate tool. Rather than having each one become its own `tmutil` program.
+Like the main `tendermint` binary, this tool could use a subcommand structure
+so that operators would not need to download many different programs, but doing
+this would isolate the specialized, dangerous, and non-production ready code
+from the main Tendermint release.
+
+Advantages:
+
+- A separate `tmutil` program could be released on a separate schedule from the
+  main Tendermint core, allowing us to quickly provide helper tools to
```

Comment:
if it's released off the same repo, it might not have a meaningfully different versioning scheme (this is fine), but it'd be easier to do "go get".

Having done something similar (e.g. managed releases for a utility CLI package related to a core service version,) the pressure moves from cutting more releases of the core to making sure the cli is compatible for nodes to an unreasonable degree.

Reply:
> if it's released off the same repo, it might not have a meaningfully different versioning scheme (this is fine), but it'd be easier to do "go get".

It's fine to host multiple modules in the same repository if we think that's valuable (the `go` tool understands namespaced tags). There are some downsides to that approach vs. a separate repo, but most of them can be papered over relatively easily. (The new "workspaces" tooling in Go 1.18 can also ease some of the headaches for us as the maintainers).